### PR TITLE
188/ChangeImmutableImplementationToWorkProperlyWithReactQuery

### PR DESCRIPTION
**Pull request by alvrba**

### CODE REVIEW

File: src/hooks/item.test.ts [hunk 1]
```diff
@@ -474,52 +488,58 @@ describe('Items Hooks', () => {
...
       const response = ITEMS;
-      const ids = response.map(({ id }) => id);
+      const ids: string[] = response.map(({ id }) => id).toArray();
       const hook = () => hooks.useItems(ids);
       const route = `/${buildGetItemsRoute(ids)}`;
       const endpoints = [{ route, response }];
       const { data } = await mockHook({ endpoints, hook, wrapper });
 
-      expect((data as Record<Item>).toJS()).toEqual(response);
+      expect((data as RecordOf<Item>).toJS()).toEqual(response.toJS());
```
Comment (pyphilia) [inline src/hooks/item.test.ts:525]: Avoid toJS().

File: src/hooks/item.test.ts [hunk 2]
```diff
@@ -543,7 +563,7 @@ describe('Items Hooks', () => {
   });
 
   describe('useItemLogin', () => {
-    const response = ITEMS[0];
+    const response = ITEMS.toArray()[0];
```
Comment (pyphilia) [inline src/hooks/item.test.ts:566]: `first()`

File: src/hooks/item.test.ts [hunk 3]
```diff
@@ -553,9 +573,9 @@ describe('Items Hooks', () => {
       const endpoints = [{ route, response }];
       const { data } = await mockHook({ endpoints, hook, wrapper });
 
-      expect((data as Record<ItemLogin>).toJS()).toEqual(response);
+      expect((data as RecordOf<ItemLogin>).toJS()).toEqual(response.toJS());
```
Comment (pyphilia) [inline src/hooks/item.test.ts:576]: Avoid toJS().

File: src/hooks/item.test.ts [hunk 4]
```diff
@@ -596,7 +616,7 @@ describe('Items Hooks', () => {
 
   describe('useFileContent', () => {
     const response = FILE_RESPONSE;
-    const { id } = ITEMS[0];
+    const { id } = ITEMS.toArray()[0];
```
Comment (pyphilia) [inline src/hooks/item.test.ts:619]: `first()`

File: src/hooks/item.test.ts [hunk 5]
```diff
@@ -695,14 +715,16 @@ describe('Items Hooks', () => {
       const response = ITEMS;
       const endpoints = [{ route, response }];
       const { data } = await mockHook({ endpoints, hook, wrapper });
-      expect((data as Record<Item>).toJS()).toEqual(response);
+      expect((data as RecordOf<Item>).toJS()).toEqual(response.toJS());
```
Comment (pyphilia) [inline src/hooks/item.test.ts:718]: Avoid toJS().

File: src/hooks/item.test.ts [hunk 6]
```diff
@@ -738,9 +760,9 @@ describe('Items Hooks', () => {
       const endpoints = [{ route, response }];
       const { data } = await mockHook({ endpoints, hook, wrapper });
 
-      expect((data as List<Item>).toJS()).toEqual(response);
+      expect((data as List<RecordOf<Item>>).toJS()).toEqual(response.toJS());
```
Comment (pyphilia) [inline src/hooks/item.test.ts:763]: Avoid toJS().

File: src/hooks/item.test.ts [hunk 7]
```diff
@@ -781,7 +803,7 @@ describe('Items Hooks', () => {
   });
 
   describe('useItemThumbnail', () => {
-    const item = ITEMS[0];
+    const item = ITEMS.toArray()[0];
```
Comment (pyphilia) [inline src/hooks/item.test.ts:806]: `first()`

File: src/hooks/item.ts [hunk 8]
```diff
@@ -183,18 +186,18 @@ export default (
 
       const { data: currentMember } = useCurrentMember();
       itemWsHooks?.useSharedItemsUpdates(
-        getUpdates ? (currentMember as Record<Member>)?.get('id') : null,
+        getUpdates ? (currentMember as RecordOf<Member>)?.id : null,
```
Comment (pyphilia) [inline src/hooks/item.ts:189]: I think you can define useCurrentMember return type to void casting the type. I think the secret is to define the type as follow:
```
return useQuery({
        queryKey: buildAppContextKey(itemId),
        queryFn: (): Promise<RecordOf<AppContext>> => {
          return Api.getContext({
            itemId,
            token,
            apiHost,
          }).then((data) => AppContextRecord(data));
        },
```
Basile did it a lot for apps-query-client, you might ask him.

File: src/hooks/item.ts [hunk 9]
```diff
@@ -241,18 +244,27 @@ export default (
...
         enabled: ids && Boolean(ids.length) && ids.every((id) => Boolean(id)),
-        ...defaultQueryOptions,
+        retry: false,
+        staleTime: 0,
+        cacheTime: 1000 * 60 * 5,
+        keepPreviousData: false,
+        refetchOnMount: false,
+        refetchOnWindowFocus: false,
+        notifyOnChangeProps: 'tracked',
+        isDataEqual: (oldData, newData) => isDataEqual(oldData, newData),
```
Comment (pyphilia) [inline src/hooks/item.ts:267]: Remove??

File: src/hooks/item.ts [hunk 10]
```diff
@@ -314,15 +328,15 @@ export default (
...
           );
         },
-        onSuccess: async (items: List<Item>) => {
+        onSuccess: async (items: List<RecordOf<Item>>) => {
           // save items in their own key
           // eslint-disable-next-line no-unused-expressions
           items?.forEach(async (item) => {
             const { id } = item;
-            queryClient.setQueryData(buildItemKey(id), Map(item));
+            queryClient.setQueryData(buildItemKey(id), item);
```
Comment (pyphilia) [inline src/hooks/item.ts:339]: Shouldn't this be a record of Item?

File: src/hooks/itemTag.ts [hunk 11]
```diff
@@ -9,14 +9,16 @@ import {
   TAGS_KEY,
 } from '../config/keys';
 import { CONSTANT_KEY_CACHE_TIME_MILLISECONDS } from '../config/constants';
+import { convertJs } from '../utils/util';
+//import { TaggedTemplateExpression } from '@babel/types';
```
Comment (pyphilia) [inline src/hooks/itemTag.ts:13]: Remove.

File: src/hooks/itemTag.ts [hunk 12]
```diff
@@ -41,16 +43,18 @@ export default (queryConfig: QueryClientConfig, queryClient: QueryClient) => {
         if (!ids) {
           throw new UndefinedArgument();
         }
-        return Api.getItemsTags(ids, queryConfig).then((data) => List(data));
+        return Api.getItemsTags(ids, queryConfig).then((data) =>
+          convertJs(data),
+        );
       },
-      onSuccess: async (tags) => {
+      onSuccess: async (tags: List<RecordOf<ItemTag>[]>) => {
```
Comment (pyphilia) [inline src/hooks/itemTag.ts:50]: Maybe this should be a list of list? I think it'll become confusing if we have a normal array in the middle.

File: src/hooks/membership.ts [hunk 13]
```diff
@@ -42,14 +43,15 @@ export default (
...
-            List(data[0]),
+            convertJs(data[0]),
           );
         },
         enabled: Boolean(id),
         ...defaultQueryOptions,
       });
     },
 
+    //REVISAR TEMA DEL ARRAY - ARRAY DE ARRAYS
```
Comment (pyphilia) [inline src/hooks/membership.ts:54]: Indeed let's use List of List ;)

File: src/queryClient.ts [hunk 14]
```diff
@@ -75,6 +76,8 @@ export default (config: Partial<QueryClientConfig>) => {
       keepPreviousData: false,
       refetchOnMount: false,
       refetchOnWindowFocus: false,
+      notifyOnChangeProps: 'tracked',
+      isDataEqual: (oldData, newData) => isDataEqual(oldData, newData),
```
Comment (pyphilia) [inline src/queryClient.ts:80]: Only `isDataEqual`

File: src/ws/hooks/item.test.ts [hunk 15]
```diff
@@ -28,15 +29,17 @@ describe('Ws Item Hooks', () => {
   });
 
   describe('useItemUpdates', () => {
-    const item = ITEMS[0];
+    const item = ITEMS.toArray()[0];
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:32]: `first()`

File: src/ws/hooks/item.test.ts [hunk 16]
```diff
@@ -28,15 +29,17 @@ describe('Ws Item Hooks', () => {
...
     const itemId = item.id;
     const itemKey = buildItemKey(itemId);
     const channel = { name: itemId, topic: TOPICS.ITEM };
-    const newItem = { ...item, description: 'new description' };
+    var changedItem = item.toJS();
+    changedItem = { ...changedItem, description: 'new description' };
+    const newItem: RecordOf<Item> = convertJs(changedItem);
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:38]: I think you can just update the value with `update` so you don't need to cast it twice.

File: src/ws/hooks/item.test.ts [hunk 17]
```diff
@@ -47,13 +50,13 @@ describe('Ws Item Hooks', () => {
 
       getHandlerByChannel(handlers, channel)?.handler(itemEvent);
 
-      expect(queryClient.getQueryData<Record<Item>>(itemKey)?.toJS()).toEqual(
-        newItem,
+      expect(queryClient.getQueryData<RecordOf<Item>>(itemKey)?.toJS()).toEqual(
+        newItem.toJS(),
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:54]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 18]
```diff
@@ -65,12 +68,12 @@ describe('Ws Item Hooks', () => {
       getHandlerByChannel(handlers, channel)?.handler(itemEvent);
 
       expect(
-        queryClient.getQueryData<Record<Item>>(itemKey)?.toJS(),
+        queryClient.getQueryData<RecordOf<Item>>(itemKey)?.toJS(),
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:71]: Since it's falsy you probably don't need to cast it?

File: src/ws/hooks/item.test.ts [hunk 19]
```diff
@@ -81,24 +84,24 @@ describe('Ws Item Hooks', () => {
 
       getHandlerByChannel(handlers, channel)?.handler(itemEvent);
 
-      expect(queryClient.getQueryData<Record<Item>>(itemKey)?.toJS()).toEqual(
-        item,
+      expect(queryClient.getQueryData<RecordOf<Item>>(itemKey)?.toJS()).toEqual(
+        item.toJS(),
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:88]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 20]
```diff
@@ -81,24 +84,24 @@ describe('Ws Item Hooks', () => {
...
       );
     });
   });
 
   describe('useChildrenUpdates', () => {
     // we need to use a different id for the channel to avoid handlers collision
-    const parent = ITEMS[1];
+    const parent = ITEMS.toArray()[1];
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:95]: `.get(1)`

File: src/ws/hooks/item.test.ts [hunk 21]
```diff
@@ -81,24 +84,24 @@ describe('Ws Item Hooks', () => {
...
     const parentId = parent.id;
     const childrenKey = buildItemChildrenKey(parentId);
     const channel = { name: parentId, topic: TOPICS.ITEM };
-    const targetItem = ITEMS[2];
+    const targetItem = ITEMS.toArray()[2];
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:99]: `get(2)`

File: src/ws/hooks/item.test.ts [hunk 22]
```diff
@@ -81,24 +84,24 @@ describe('Ws Item Hooks', () => {
...
     const targetItemKey = buildItemKey(targetItem.id);
     const hook = () => hooks.useChildrenUpdates(parentId);
 
     it(`Receive create child`, async () => {
-      queryClient.setQueryData(childrenKey, List([ITEMS[3]]));
+      queryClient.setQueryData(childrenKey, List([ITEMS.toArray()[3]]));
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:104]: `.get(3)`
And probably we want a list of list.

File: src/ws/hooks/item.test.ts [hunk 23]
```diff
@@ -112,17 +115,19 @@ describe('Ws Item Hooks', () => {
       // check children key contains new item
       expect(
         queryClient.getQueryData<List<Item>>(childrenKey)?.toJS(),
-      ).toContainEqual(targetItem);
+      ).toContainEqual(targetItem.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:118]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 24]
```diff
@@ -112,17 +115,19 @@ describe('Ws Item Hooks', () => {
...
       // check new item key
       expect(
-        queryClient.getQueryData<Record<Item>>(targetItemKey)?.toJS(),
-      ).toEqual(targetItem);
+        queryClient.getQueryData<RecordOf<Item>>(targetItemKey)?.toJS(),
+      ).toEqual(targetItem.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:122]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 25]
```diff
@@ -112,17 +115,19 @@ describe('Ws Item Hooks', () => {
...
     });
 
     it(`Receive update child`, async () => {
-      const updatedItem = { ...targetItem, description: 'new description' };
-      queryClient.setQueryData(targetItemKey, Map(targetItem));
-      queryClient.setQueryData(childrenKey, List(ITEMS));
+      var changedItem = targetItem.toJS();
+      changedItem = { ...changedItem, description: 'new description' };
+      const updatedItem: RecordOf<Item> = convertJs(changedItem);
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:128]: Same as above, use `update`.

File: src/ws/hooks/item.test.ts [hunk 26]
```diff
@@ -135,17 +140,19 @@ describe('Ws Item Hooks', () => {
 
       // check new item key content
       expect(
-        queryClient.getQueryData<Record<Item>>(targetItemKey)?.toJS(),
-      ).toEqual(updatedItem);
+        queryClient.getQueryData<RecordOf<Item>>(targetItemKey)?.toJS(),
+      ).toEqual(updatedItem.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:144]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 27]
```diff
@@ -135,17 +140,19 @@ describe('Ws Item Hooks', () => {
...
       // check children key contains newly item
-      const own = queryClient.getQueryData<List<Item>>(childrenKey)?.toJS();
-      expect(own).toContainEqual(updatedItem);
-      expect(own?.length).toBe(ITEMS.length);
+      const own = queryClient
+        .getQueryData<List<RecordOf<Item>>>(childrenKey)
+        ?.toJS();
+      expect(own).toContainEqual(updatedItem.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:149]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 28]
```diff
@@ -176,20 +183,20 @@ describe('Ws Item Hooks', () => {
       getHandlerByChannel(handlers, channel)?.handler(itemEvent);
 
       expect(
-        queryClient.getQueryData<Record<Item>>(childrenKey)?.toJS(),
-      ).toEqual(ITEMS);
+        queryClient.getQueryData<RecordOf<Item>>(childrenKey)?.toJS(),
+      ).toEqual(ITEMS.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:187]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 29]
```diff
@@ -202,18 +209,20 @@ describe('Ws Item Hooks', () => {
...
       expect(
-        queryClient.getQueryData<List<Item>>(OWN_ITEMS_KEY)?.toJS(),
-      ).toContainEqual(item);
+        queryClient.getQueryData<List<RecordOf<Item>>>(OWN_ITEMS_KEY)?.toJS(),
+      ).toContainEqual(item.toJS());
       // check new item key
-      expect(queryClient.getQueryData<Record<Item>>(itemKey)?.toJS()).toEqual(
-        item,
+      expect(queryClient.getQueryData<RecordOf<Item>>(itemKey)?.toJS()).toEqual(
+        item.toJS(),
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:216]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 30]
```diff
@@ -225,20 +234,20 @@ describe('Ws Item Hooks', () => {
...
+        updatedItem.toJS(),
       );
       // check children key contains newly item
       const children = queryClient
-        .getQueryData<List<Item>>(OWN_ITEMS_KEY)
+        .getQueryData<List<RecordOf<Item>>>(OWN_ITEMS_KEY)
         ?.toJS();
-      expect(children).toContainEqual(updatedItem);
-      expect(children?.length).toBe(ITEMS.length);
+      expect(children).toContainEqual(updatedItem.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:244]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 31]
```diff
@@ -270,21 +279,21 @@ describe('Ws Item Hooks', () => {
       getHandlerByChannel(handlers, channel)?.handler(itemEvent);
 
       expect(
-        queryClient.getQueryData<List<Item>>(OWN_ITEMS_KEY)?.toJS(),
-      ).toEqual(ITEMS);
+        queryClient.getQueryData<List<RecordOf<Item>>>(OWN_ITEMS_KEY)?.toJS(),
+      ).toEqual(ITEMS.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:283]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 32]
```diff
@@ -270,21 +279,21 @@ describe('Ws Item Hooks', () => {
...
     });
   });
 
   describe('useSharedItemsUpdates', () => {
     // we need to use a different id to avoid handler collision
-    const item = ITEMS[1];
+    const item = ITEMS.toArray()[1];
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:289]: `.get(1)`

File: src/ws/hooks/item.test.ts [hunk 33]
```diff
@@ -297,18 +306,22 @@ describe('Ws Item Hooks', () => {
...
-      ).toContainEqual(item);
+        queryClient
+          .getQueryData<List<RecordOf<Item>>>(SHARED_ITEMS_KEY)
+          ?.toJS(),
+      ).toContainEqual(item.toJS());
       // check new item key
-      expect(queryClient.getQueryData<Record<Item>>(itemKey)?.toJS()).toEqual(
-        item,
+      expect(queryClient.getQueryData<RecordOf<Item>>(itemKey)?.toJS()).toEqual(
+        item.toJS(),
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:315]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 34]
```diff
@@ -297,18 +306,22 @@ describe('Ws Item Hooks', () => {
...
       );
     });
 
     it(`Receive update child`, async () => {
-      const updatedItem = { ...item, description: 'new description' };
-      queryClient.setQueryData(itemKey, Map(item));
-      queryClient.setQueryData(SHARED_ITEMS_KEY, List(ITEMS));
+      var changedItem = item.toJS();
+      changedItem = { ...changedItem, description: 'new description' };
+      const updatedItem: RecordOf<Item> = convertJs(changedItem);
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:322]: Use `update`.

File: src/ws/hooks/item.test.ts [hunk 35]
```diff
@@ -320,20 +333,20 @@ describe('Ws Item Hooks', () => {
       getHandlerByChannel(handlers, channel)?.handler(itemEvent);
 
       // check new item key content
-      expect(queryClient.getQueryData<Record<Item>>(itemKey)?.toJS()).toEqual(
-        updatedItem,
+      expect(queryClient.getQueryData<RecordOf<Item>>(itemKey)?.toJS()).toEqual(
+        updatedItem.toJS(),
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:337]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 36]
```diff
@@ -320,20 +333,20 @@ describe('Ws Item Hooks', () => {
...
       );
       // check children key contains newly item
       const shared = queryClient
-        .getQueryData<List<Item>>(SHARED_ITEMS_KEY)
+        .getQueryData<List<RecordOf<Item>>>(SHARED_ITEMS_KEY)
         ?.toJS();
-      expect(shared).toContainEqual(updatedItem);
-      expect(shared?.length).toBe(ITEMS.length);
+      expect(shared).toContainEqual(updatedItem.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:343]: Avoid toJS().

File: src/ws/hooks/item.test.ts [hunk 37]
```diff
@@ -365,8 +378,10 @@ describe('Ws Item Hooks', () => {
       getHandlerByChannel(handlers, channel)?.handler(itemEvent);
 
       expect(
-        queryClient.getQueryData<List<Item>>(SHARED_ITEMS_KEY)?.toJS(),
-      ).toEqual(ITEMS);
+        queryClient
+          .getQueryData<List<RecordOf<Item>>>(SHARED_ITEMS_KEY)
+          ?.toJS(),
+      ).toEqual(ITEMS.toJS());
```
Comment (pyphilia) [inline src/ws/hooks/item.test.ts:384]: Avoid toJS().

**Review by pyphilia — CHANGES_REQUESTED** [review]
Nice changes, I'm looking forward it! Indeed it will be a create idea to create types for `Record<x>`. You could also export it, so that once we change the interfaces to typescript, we can reuse them.
It's often the same comments, I'll let you refactor that! It's cool it seems to be cleaner than before^^

File: src/hooks/item.ts (hunk 10 above)
Comment (alvrba) [reply src/hooks/item.ts:339]: It is already a Record<Item> (the convertJs function already converts it to Record)

File: src/utils/util.ts [hunk 38]
```diff
@@ -13,3 +15,31 @@ export const getHostname = () => {
...
+
+  return new Factory(Seq<any>(data).map(convertJs));
+}
+
+export const isDataEqual = (
+  oldData:
+    | RecordOf<any>
+    | List<RecordOf<any>>
+    | List<RecordOf<any>>[]
+    | undefined,
```
[the remark this answers is not in the document]
Comment (alvrba) [reply src/utils/util.ts:38]: The types are defined in the next lines, prettier has done this structure

File: src/types.ts [hunk 39]
```diff
@@ -16,6 +18,15 @@ export type QueryClientConfig = {
     refetchOnWindowFocus?: boolean;
     keepPreviousData?: boolean;
     refetchOnMount?: boolean;
+    notifyOnChangeProps?: any;
+    isDataEqual: (
+      oldData:
```
[the remark this answers is not in the document]
Comment (alvrba) [reply src/types.ts:23]: The types are defined in the next lines, prettier has done this structure

File: src/hooks/item.ts [hunk 40]
```diff
@@ -237,18 +238,25 @@ export default (
...
+          }
+          if (ids.length === 1) {
+            return Api.getItem(ids[0], queryConfig).then((data) =>
+              convertJs([data]),
+            );
+          } else {
+            return Api.getItems(ids, queryConfig).then((data) =>
+              convertJs(data),
+            );
+          }
```
Comment (pyphilia) [inline src/hooks/item.ts:253]: Unnecessary else clause, since you return in the previous if .

File: src/hooks/itemLike.test.ts [hunk 41]
```diff
@@ -2,7 +2,7 @@
 import nock from 'nock';
 import Cookies from 'js-cookie';
 import { StatusCodes } from 'http-status-codes';
-import { List } from 'immutable';
+//import { List } from 'immutable';
```
Comment (pyphilia) [inline src/hooks/itemLike.test.ts:5]: Remove?

File: src/types.ts (hunk 39 above)
[the remark this answers is not in the document]
Comment (pyphilia) [reply src/types.ts:23]: Sorry it was not clear, my point was to change with the following:
```
oldData?: | RecordOf<any>
        | List<RecordOf<any>>
        | List<List<RecordOf<any>>>
```

File: src/utils/util.ts [hunk 42]
```diff
@@ -13,3 +15,31 @@ export const getHostname = () => {
...
+    | List<RecordOf<any>>
+    | List<List<RecordOf<any>>>
+    | undefined,
+  newData: RecordOf<any> | List<RecordOf<any>> | List<List<RecordOf<any>>>,
+): boolean => {
+  if (is(oldData, newData)) {
+    return true;
+  }
+  return false;
+};
```
Comment (pyphilia) [inline src/utils/util.ts:45]: Refactor with:
```
return is(oldData, newData)
```

File: src/ws/hooks/chat.ts [hunk 43]
```diff
@@ -1,8 +1,10 @@
-import { Record } from 'immutable';
+import { List, RecordOf } from 'immutable';
 import { useEffect } from 'react';
 import { QueryClient } from 'react-query';
+//import { MESSAGE_IDS } from '../../../test/constants';
```
Comment (pyphilia) [inline src/ws/hooks/chat.ts:4]: Remove.

**Review by pyphilia — APPROVED** [review]
It looks really good, I love this refactor. I've added minor comments and it is good to merge!

File: src/types.ts (hunk 39 above)
[the remark this answers is not in the document]
Comment (alvrba) [reply src/types.ts:23]: Screenshot 2022-08-12 at 10 54 48
I defined it like in the react query documentation, it seems not to be the same the ? and undefined. 

This is from SO:
? isn't shorthand for string | undefined. It's an optional parameter. Meaning - you can either pass the expected value type, or not.

string | undefined, by contrast, means you must explicitly pass a value that is either a string or undefined. The value may be undefined, but the parameter still exists.

**Pull request merged**